component/xload/comp/splitter.go

```diff
@@ -0,0 +1,201 @@
+	d.GetThreadPool().Start()
+}
+
+func (d *downloadSplitter) Stop() {
+	log.Debug("downloadSplitter::Stop : stop download splitter for %s", d.path)
+	if d.GetThreadPool() != nil {
+		d.GetThreadPool().Stop()
+	}
+	d.GetNext().Stop()
+}
+
+// download data in chunks and then write to the local file
+func (d *downloadSplitter) Process(item *common.WorkItem) (int, error) {
+	var err error
+
+	log.Debug("downloadSplitter::Process : Splitting data for %s", item.Path)
+	if len(item.Path) == 0 {
+		return 0, nil
+	}
+
+	numBlocks := ((item.DataLen - 1) / d.blockSize) + 1
+	offset := int64(0)
+
+	// TODO:: xload : should we delete the file if it already exists
+	// TODO:: xload : what should be the flags and mode and should we allocate the full size to the file
+	item.FileHandle, err = os.OpenFile(filepath.Join(d.path, item.Path), os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		// create file
+		return -1, fmt.Errorf("failed to open file %s [%v]", item.Path, err)
+	}
+
+	defer item.FileHandle.Close()
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	responseChannel := make(chan *common.WorkItem, numBlocks)
+
+	operationSuccess := true
+	go func() {
+		defer wg.Done()
+
+		for i := 0; i < int(numBlocks); i++ {
+			respSplitItem := <-responseChannel
+			if respSplitItem.Err != nil {
+				log.Err("downloadSplitter::Process : Failed to download data for file %s", item.Path)
+				operationSuccess = false
+			} else {
+				_, err := item.FileHandle.WriteAt(respSplitItem.Block.Data, respSplitItem.Block.Offset)
+				if err != nil {
+					log.Err("downloadSplitter::Process : Failed to write data to file %s", item.Path)
+					operationSuccess = false
+				}
+			}
+
+			if respSplitItem.Block != nil {
+				log.Debug("downloadSplitter::Process : Download successful %s index %d offset %v", item.Path, respSplitItem.Block.Index, respSplitItem.Block.Offset)
+				d.blockPool.Release(respSplitItem.Block)
+			}
+		}
+	}()
+
+	for i := 0; i < int(numBlocks); i++ {
+		block := d.blockPool.Get()
+		if block == nil {
+			responseChannel <- &common.WorkItem{Err: fmt.Errorf("failed to get block from pool for file %s, offset %v", item.Path, offset)}
+		} else {
+			block.Index = i
+			block.Offset = offset
+			block.Length = int64(d.blockSize)
+
+			splitItem := &common.WorkItem{
+				CompName:        d.GetNext().GetName(),
+				Path:            item.Path,
+				DataLen:         item.DataLen,
+				FileHandle:      item.FileHandle,
+				Block:           block,
+				ResponseChannel: responseChannel,
+				Download:        true,
+			}
+			log.Debug("downloadSplitter::Process : Scheduling download for %s offset %v", item.Path, offset)
+			d.GetNext().GetThreadPool().Schedule(splitItem)
+		}
+
+		offset += int64(d.blockSize)
+	}
+
+	wg.Wait()
+	err = item.FileHandle.Truncate(int64(item.DataLen))
+	if err != nil {
+		log.Err("downloadSplitter::Process : Failed to truncate file %s [%s]", item.Path, err.Error())
+		operationSuccess = false
+	}
+
+	if !operationSuccess {
```

Comment:
Shall we update the last modified time of the local file to the one mentioned in container? Not sure but applications might take decisions based on that.